### PR TITLE
Add numpy backend tests for Gaussian

### PR DESCRIPTION
**Pull request by fehiepsi**
Follow up  #303, this PR adds numpy backend tests for gaussian.

### Tasks

- [x] fix eager_subs_affine test; currently, the numpy test failed due to torch tensor is created during the process...
- [x] fix integrate tests

### CODE REVIEW

File: funsor/affine.py [hunk 1]
```diff
@@ -135,10 +134,12 @@ def extract_affine(fn):
         ``(coefficient, eqn)`` pair in einsum form.
     :rtype: tuple
     """
+    # FIXME: does this make sense?
+    prototype = fn.terms[1].data
```
Comment (fehiepsi) [inline funsor/affine.py]: @fritzo I need prototype value for the below `ops.new_zeros`. This line passes all gaussian tests but does not work with general `fn` (of course). Do you have an idea to resolve this issue? It seems to me that `extract_affine` is only used in Gaussian. Should we make it a method of gaussian (to use backend information from `info_vec` or `precision`)?

File: funsor/cnf.py [hunk 2]
```diff
@@ -231,14 +231,14 @@ def eager_contraction_to_binary(red_op, bin_op, reduced_vars, lhs, rhs):
 def eager_contraction_tensor(red_op, bin_op, reduced_vars, *terms):
     if not all(term.dtype == "real" for term in terms):
         raise NotImplementedError('TODO')
-    return _eager_contract_tensors(reduced_vars, terms, backend="torch")
+    return _eager_contract_tensors(reduced_vars, terms, backend=terms[0].backend)
```
Comment (fehiepsi) [inline funsor/cnf.py]: `Tensor.backend` is a string of `"torch"` or `"numpy"`.

File: funsor/cnf.py [hunk 3]
```diff
@@ -231,14 +231,14 @@ def eager_contraction_to_binary(red_op, bin_op, reduced_vars, lhs, rhs):
...
 
 
 @eager.register(Contraction, ops.LogAddExpOp, ops.AddOp, frozenset, Tensor, Tensor)
 def eager_contraction_tensor(red_op, bin_op, reduced_vars, *terms):
     if not all(term.dtype == "real" for term in terms):
         raise NotImplementedError('TODO')
-    return _eager_contract_tensors(reduced_vars, terms, backend="pyro.ops.einsum.torch_log")
+    return _eager_contract_tensors(reduced_vars, terms, backend=terms[0].logsumexp_backend)
```
Comment (fehiepsi) [inline funsor/cnf.py]: Tensor.logsumexp_backend is either `pyro.ops.einsum.torch_log` or `funsor.numpy`

File: funsor/numpy.py [hunk 4]
```diff
@@ -7,6 +7,45 @@
 from funsor.util import quote
 
 
+def einsum(equation, *operands):
```
Comment (fehiepsi) [inline funsor/numpy.py]: adapted from Pyro [torch_log](https://github.com/pyro-ppl/pyro/blob/dev/pyro/ops/einsum/torch_log.py)

File: funsor/testing.py [hunk 5]
```diff
@@ -241,18 +241,18 @@ def random_tensor(inputs, output=reals(), backend="torch"):
...
     batch_shape = tuple(d.dtype for d in inputs.values() if d.dtype != 'real')
     event_shape = (sum(d.num_elements for d in inputs.values() if d.dtype == 'real'),)
-    prec_sqrt = torch.randn(batch_shape + event_shape + event_shape)
-    precision = torch.matmul(prec_sqrt, prec_sqrt.transpose(-1, -2))
-    precision = precision + 0.05 * torch.eye(event_shape[0])
-    loc = torch.randn(batch_shape + event_shape)
-    info_vec = precision.matmul(loc.unsqueeze(-1)).squeeze(-1)
+    prec_sqrt = randn(batch_shape + event_shape + event_shape, backend)
+    precision = ops.matmul(prec_sqrt, ops.transpose(prec_sqrt, -1, -2))
+    precision = precision + 0.5 * ops.new_eye(precision, event_shape[:1])
```
Comment (fehiepsi) [inline funsor/testing.py]: Here I change `0.05` to `0.5` to pass numpy tests. With `0.05`, even with torch backend, gaussian integrate tests will fail with some random seed.

File: funsor/affine.py (hunk 1 above)
Comment (fehiepsi) [reply funsor/affine.py]: A simple solution is to introduce an arg `backend` here, and use Tensor(...).backend to dispatch.

File: funsor/cnf.py (hunk 2 above)
Comment (fritzo) [reply funsor/cnf.py]: I think it is better to put this in a new file funsor/einsum/numpy_log.py. Note the `opt_einsum` requirements are a little subtle: they use a `tensordot`, an `einsum` and a `transpose` implementation. Since you've already implemented a logsumexp `einsum`, you can create a `tensordot` via
```py
from pyro.ops.einsum.util import Tensordot
tensordot = Tensordot(einsum)
```

File: funsor/numpy.py (hunk 4 above)
Comment (fritzo) [reply funsor/numpy.py]: As I mentioned above, this should go in a separate file like funsor/einsum/numpy_log.py, in particular because we might want to add a linear (sum-product) or a tropical (max,sum) version in the future, and we'll want to set e.g. `backend=funsor.einsum.numpy_tropical` for those cases. Note these are not user-facing; they are only consumed by `opt_einsum`.

File: funsor/numpy.py [hunk 6]
```diff
@@ -7,6 +7,45 @@
...
+    """
+    Log-sum-exp implementation of einsum.
+    """
+    # rename symbols to support PyTorch 0.4.1 and earlier,
```
Comment (fritzo) [inline funsor/numpy.py]: nit: Does this still apply to NumPy?

File: funsor/pyro/convert.py [hunk 7]
```diff
@@ -82,12 +82,12 @@ def tensor_to_funsor(tensor, event_inputs=(), event_output=0, dtype="real"):
 
 def funsor_to_tensor(funsor_, ndims, event_inputs=()):
     """
-    Convert a :class:`funsor.torch.Tensor` to a :class:`torch.Tensor` .
+    Convert a :class:`funsor.tensor.Tensor` to a :class:`torch.Tensor` .
```
Comment (fritzo) [inline funsor/pyro/convert.py]: I guess we'll eventually want to use this same helper for `np.array`, so we can convert to a raw `numeric_array` of underlying type?

File: funsor/tensor.py [hunk 8]
```diff
@@ -381,6 +381,20 @@ def materialize(self, x):
...
 
+    @property
+    def backend(self):
+        if torch.is_tensor(self.data):
+            return "torch"
+        else:
+            return "numpy"
+
+    @property
+    def logsumexp_backend(self):
```
Comment (fritzo) [inline funsor/tensor.py]: nit: I think this is better as a global dict accessible to `_eager_contract_tensors()`:
```py
# in funsor/cnf.py
BACKEND_TO_LOGSUMEXP_BACKEND = {
    "torch": "pyro.ops.einsum.torch_log",
    "numpy": "funsor.einsum.numpy_log",
}

@eager.register(Contraction, ops.LogAddExpOp, ops.AddOp, frozenset, Tensor, Tensor)
def eager_contraction_tensor(red_op, bin_op, reduced_vars, *terms):
    if not all(term.dtype == "real" for term in terms):
        raise NotImplementedError('TODO')
    backend = BACKEND_TO_LOGSUMEXP_BACKEND[terms[0].backend]
    return _eager_contract_tensors(reduced_vars, terms, backend=backend)
```

File: funsor/affine.py (hunk 1 above)
Comment (fritzo) [reply funsor/affine.py]: This seems reasonable, and preferable to a global option.

File: funsor/pyro/convert.py (hunk 7 above)
Comment (fehiepsi) [reply funsor/pyro/convert.py]: Yes, I think we should move these utilities to e.g. `tensor.py` file. I would like to defer this to the relevant PR (where we can discuss which file those utilities should be in).

File: funsor/affine.py [hunk 9]
```diff
@@ -135,10 +135,12 @@ def extract_affine(fn):
         ``(coefficient, eqn)`` pair in einsum form.
     :rtype: tuple
     """
+    # FIXME: does this make sense?
+    prototype = prototype if prototype is not None else torch.tensor([])
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply funsor/affine.py]: I think having a global setting is a good idea. Currently, we don't need a global setting for remaining dispatching code (except `arange` and `materialize`, which I have converted them to Tensor methods), so I made a temporary implementation to only use this global setting in Gaussian affine subs.

File: funsor/numpy.py (hunk 6 above)
Comment (fehiepsi) [reply funsor/numpy.py]: It applies to NumPy, but in JAX we don't need this.

File: funsor/tensor.py (hunk 8 above)
Comment (fehiepsi) [reply funsor/tensor.py]: Thanks!

File: funsor/einsum/numpy_log.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import numpy as np
```
Comment (fritzo) [inline funsor/einsum/numpy_log.py]: I believe [opt_einsum requires](https://optimized-einsum.readthedocs.io/en/stable/backends.html) also `tensordot` and `transpose` implementations, which you can get via
```py
from pyro.ops.einsum.util import Tensordot

tensordot = Tensordot(einsum)
transpose = np.transpose
```
These may not be exercised in existing tests, but I believe they will be required for more complex contractions in e.g. NumPyro's enumeration machinery.

**Comment by fehiepsi** [comment]
Yes, I have made a separate issue for `test_einsum` at https://github.com/pyro-ppl/funsor/issues/1#issue-410457128. Thanks for your suggestion!!
